### PR TITLE
Fix duplicate layers in exported tar

### PR DESCRIPTION
**Pull request by tonistiigi**
`docker save` stores the image layer's paths using random IDs to ensure compatibility with versions `<1.10`. This means that when same layer is exported multiple times as part of a single `docker save` operation, duplicate layers will be put in the tarball. This is especially bad when saving whole parent chains where almost all layers always match. This PR adds a cache based on layer diffID, and creates a symlink if same layer is used again instead of making a full copy.

Signed-off-by: Tonis Tiigi tonistiigi@gmail.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
I think there's an issue for this somewhere

Also ping @aaronlehmann

**Comment by tonistiigi** [comment]
> I think there's an issue for this somewhere

It's been an issue for people saving the parent chain to transport the build cache.

File: image/tarexport/save.go [hunk 1]
```diff
@@ -297,36 +299,47 @@ func (s *saveSession) saveLayer(id layer.ChainID, legacyImg image.V1Image, creat
...
-	if err != nil {
-		return distribution.Descriptor{}, err
-	}
-	defer arch.Close()
+	if oldPath, exists := s.diffIDPaths[l.DiffID()]; exists {
+		relPath, err := filepath.Rel(layerPath, oldPath)
+		if err != nil {
+			return distribution.Descriptor{}, err
+		}
+		os.Symlink(relPath, layerPath)
```
Comment (aaronlehmann) [inline image/tarexport/save.go]: I think `savedLayers` should also be updated in this case.

Comment (tonistiigi) [inline image/tarexport/save.go]: updated

**Comment by aaronlehmann** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**